src/android/mod.rs

```diff
@@ -108,10 +109,10 @@ pub unsafe fn android_setup(
   };
 
   looper
-    .add_fd_with_callback(MAIN_PIPE[0], FdEvent::INPUT, move |_| {
+    .add_fd_with_callback(MAIN_PIPE[0].as_fd(), FdEvent::INPUT, move |fd, _event| {
       let size = std::mem::size_of::<bool>();
       let mut wake = false;
-      if libc::read(MAIN_PIPE[0], &mut wake as *mut _ as *mut _, size) == size as libc::ssize_t {
+      if libc::read(fd.as_raw_fd(), &mut wake as *mut _ as *mut _, size) == size as libc::ssize_t {
         main_pipe.recv().is_ok()
```

Comment:
Note that the compiler points out that `main_pipe` is not `Send`, so this won't compile.

Reply:
It is probably safe to just manually implement `Send` for main_pipe, it is sent here and never accessed anywhere else. It is a workaround for now, hopefully I can find time to switch to android-activity crate instead.

Reply:
@amrbashir don't ever push such horrendous hacks to external contributors' PRs. I can't even refute your safety assessment because there is none to begin with.

This `Send` requirement stems from the `JNIEnv`, and the documentation clearly says **in bold** that they are thread-local and **cannot be shared between threads**:
https://developer.android.com/training/articles/perf-jni

(I hope this is not how you typcially handle safety-related compiler errors in `wry`/`tao` :see_no_evil:)

Reply:
In any case, the correct fix is to pass the `ThreadLooper` which does not have the `Send` bound. Looking through the `tao` implementation (for which I'll send an update-PR shortly - had that in the queue while resolving this PR first) it calls this `android_setup()` function on the thread where both the `JNIEnv` are and looper are local so this is the only correct thing to do.

Reply:
> @amrbashir don't ever push such horrendous hacks to external contributors' PRs.

Apologies

> I can't even refute your safety assessment because there is none to begin with.
> 
> This `Send` requirement stems from the `JNIEnv`, and the documentation clearly says **in bold** that they are thread-local and **cannot be shared between threads**: [developer.android.com/training/articles/perf-jni](https://developer.android.com/training/articles/perf-jni)

I know that `MainPipe` is not `Send` but it is only ever accessed on the same thread and never shared between other threads so that's why it is probably safe to just slap a `unsafe imple Send for MainPipe`. 


> (I hope this is not how you typically handle safety-related compiler errors in `wry`/`tao` 🙈)

As I said before, our current android implementation is all a hack and experimental, and we currently don't have the time to rewrite a better implementation so I myself just try to get by if possible. 


Anywyas, thanks for fixing this properly, we appreciate it.

Reply:
> but it is only ever accessed on the same thread and never shared between other threads so that's why it is probably safe to just slap a `unsafe impl Send for MainPipe` and as a good measure because I didn't trust my assessment fully, I made sure to test it on a real application which seemed to work fine.

You didn't say that _"it is only ever accessed on the same thread"_ in your original assessment. This is the kind of info that should have gone into a `// SAFETY:` comment.

> As I said before, our current android implementation is all a hack and experimental

That doesn't seem to have been stated anywhere, just that the `unsafe impl Send` was a workaround until switching to a different kind of implementation i.e. based on `android-activity`. But you're right, it doesn't take much reading of the code to realize this :wink: 

At least we managed to patch it up somewhat.